nomad/deploymentwatcher/batcher.go

```diff
@@ -0,0 +1,141 @@
+package deploymentwatcher
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	"github.com/hashicorp/nomad/nomad/structs"
+)
+
+// EvalBatcher is used to batch the creation of evaluations
+type EvalBatcher struct {
+	// batch is the batching duration
+	batch time.Duration
+
+	// raft is used to actually commit the evaluations
+	raft DeploymentRaftEndpoints
+
+	// future to be returned to callers
+	f *EvalFuture
+
+	// inCh is used to pass evaluations to the daemon process
+	inCh chan *structs.Evaluation
+
+	// ctx is used to exit the daemon batcher
+	ctx context.Context
+
+	l sync.Mutex
+}
+
+// NewEvalBatcher returns an EvalBatcher that uses the passed raft endpoints to
+// create the evaluations and exits the batcher when the passed exit channel is
+// closed.
+func NewEvalBatcher(batchDuration time.Duration, raft DeploymentRaftEndpoints, ctx context.Context) *EvalBatcher {
+	b := &EvalBatcher{
+		batch: batchDuration,
+		raft:  raft,
+		ctx:   ctx,
+		inCh:  make(chan *structs.Evaluation, 10),
+	}
+
+	go b.batcher()
+	return b
+}
+
+// CreateEval batches the creation of the evaluation and returns a future that
+// tracks the evaluations creation.
+func (b *EvalBatcher) CreateEval(e *structs.Evaluation) *EvalFuture {
+	b.l.Lock()
+	if b.f == nil {
+		b.f = NewEvalFuture()
+	}
+	b.l.Unlock()
+
+	b.inCh <- e
+	return b.f
+}
+
+// batcher is the long lived batcher goroutine
+func (b *EvalBatcher) batcher() {
+	timer := time.NewTimer(b.batch)
+	evals := make(map[string]*structs.Evaluation)
+	for {
+		select {
+		case <-b.ctx.Done():
+			timer.Stop()
+			return
+		case e := <-b.inCh:
+			if len(evals) == 0 {
+				if !timer.Stop() {
```

Comment:
You can't :( https://golang.org/pkg/time/#Timer.Reset

(Easily the worst API in go's stdlib)